src/parser/parser_v2.cpp

```diff
@@ -393,44 +393,44 @@ rule_filter_spec parse_rule_filter(
 
 } // namespace
 
-rule_spec_container parse_rules(parameter::vector &rule_array, ddwaf::ruleset_info &info,
+rule_spec_container parse_rules(parameter::vector &rule_array, base_section_info &info,
     manifest &target_manifest, std::unordered_map<std::string, std::string> &rule_data_ids,
     const object_limits &limits, rule::source_type source)
 {
     rule_spec_container rules;
-    for (const auto &rule_param : rule_array) {
+    for (unsigned i = 0; i < rule_array.size(); ++i) {
+        const auto &rule_param = rule_array[i];
         auto rule_map = static_cast<parameter::map>(rule_param);
         std::string id;
         try {
             id = at<std::string>(rule_map, "id");
             if (rules.find(id) != rules.end()) {
                 DDWAF_WARN("duplicate rule %s", id.c_str());
-                info.insert_error(id, "duplicate rule");
+                info.add_failed(id, "duplicate rule");
                 continue;
             }
 
             auto rule = parse_rule(rule_map, target_manifest, rule_data_ids, limits, source);
+            info.add_loaded(id);
             rules.emplace(std::move(id), std::move(rule));
-            info.add_loaded();
         } catch (const std::exception &e) {
-            if (!id.empty()) {
-                DDWAF_WARN("failed to parse rule '%s': %s", id.c_str(), e.what());
-                info.insert_error(id, e.what());
-            } else {
-                DDWAF_WARN("failed to parse rule: %s", e.what());
-                info.add_failed();
+            if (id.empty()) {
+                id = "index:" + std::to_string(i);
```

Comment:
It's not a major thing but maybe it would be good idea to extract a function which generates an `id` from a number. It's business logic and it's duplicated a few times on this file